### PR TITLE
Remove JSONElement from 6.x; replace with roundtrip using JSONSerialization

### PR DESCRIPTION
**Pull request by yeahphil**
This branch removes JSONElement, a recursive enum that can represent a JSON object. JSONElement was used as an intermediate representation while parsing messages. The pull request uses JSONSerialization's output as this intermediate format.

Handling Phoenix Channel messages if you want type-safe JSON decoding is surprisingly tricky. Probably you want to end up with Codable types, but without parsing the first few fields of the message to get the topic/event, it can't be routed and so you won't know what type to decode the payload to.

There are a couple ways out of this: 
1.  Partially decode the first part of the message, route it, then decode the whole message once the type is known
2. Decode the whole message, then re-encode the payload to Data for further processing

1 is actually a bit faster than 2 (at least on newer versions of iOS), but makes it difficult to provide just the payload as a Data. So it would be a doable approach, but would require a much higher surface area change in this library.

2 is the approach the 6.x branch is set up for, and which this PR continues. Using a recursive enum to decode JSON is a popular approach (https://github.com/supabase/supabase-swift also uses it in their Phoenix Channels implementation). But it does not perform well.

Benchmarks for various decoding strategies are here: https://github.com/yeahphil/CodablePerformance -- these use Supabase's AnyJSON, which has a very similar approach and performance to JSONElement.

Decode times running on a simulator on my dev machine against a large (~3.2MB) JSON object:

| Method | Runtime |
| -------- | ------- |
| [String: Any] (using JSONSerialization) | 0.029 sec |
| Codable (using JSONDecoder) | 0.039 sec |
| AnyJSON (using JSONDecoder) | 0.615 sec |

Decoding to a recursive enum for this large payload is 21x slower! That's a big enough slowdown in de-serialization to care about. (This effect might lessen and not matter for really small messages, but it's still about an order of magnitude slower for a ~30kb payload)

### CODE REVIEW

**Comment by dsrees** [comment]
> Partially decode the first part of the message, route it, then decode the whole message once the type is known

I'm curious what you envision this approach would look like

**Comment by yeahphil** [comment]
Sure -- that could look something like:

```swift
public struct MessageHeader: Codable, Equatable {
    let joinRef: String?
    let ref: String?
    let topic: String
    let event: String
    
    public init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        joinRef = try? container.decode(String?.self)
        ref = try? container.decode(String?.self)
        topic = try container.decode(String.self)
        event = try container.decode(String.self)
        
        if event == Channel.Events.reply {
            let wrapper = try container.decode(ResponseWrapper.self)
            status = wrapper.status
        } else {
            status = nil
        }
    }
}

// This is fine. JSONDecoder doesn't care that there's extra elements in the array that don't get decoded.
let header = try JSONDecoder().decode(MessageHeader.self, from: incomingMessageData)
```

On iOS 17+, for big messages, that's faster than deserializing the whole message. 

The catch is, you can't extract `Data` for just the message payload from that, so you have to pass the whole `incomingMesageData` down to subscribers, with the wrapping array and header fields.

Once the message is routed and the expected type is known, the payload can be turned into a Codable with:

```swift
public struct CodableMessage<T: Codable>: Codable {
    public let joinRef: String?
    public let ref: String?
    public let topic: String
    public let event: String
    public let payload: T
    
    public init(from decoder: Decoder) throws {
        var container = try decoder.unkeyedContainer()
        joinRef = try? container.decode(String?.self)
        ref = try? container.decode(String?.self)
        topic = try container.decode(String.self)
        event = try container.decode(String.self)
        payload = try container.decode(T.self)
    }
}

let someTypedPayload = try JSONDecoder().decode(CodableMessage<SomeCodableType>.self, from: incomingMessageData).payload
```

You can see that I got pretty far down this path before switching to just round tripping through `JSONSerialization`. In a lot of ways it's a better approach! But the catch is, either the library user has to be aware of the Phoenix Channels message wrapper, or everything that returns a `Message` needs to be reworked to take a `Codable.Type`, something like:
```swift
func on<T: Codable>(_ event: String, type: T.Type, callback: @escaping (T) -> Void)
```
Then you still might want a version that calls back with a `[String: Any]`, otherwise existing users will have to provide `Codable` types for everything in order to upgrade.

For an existing library with a long history, providing `Data` of the payload provides an easier upgrade path and the change is way less surface area. But it has some tradeoffs.

File: Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift [hunk 1]
```diff
@@ -91,37 +91,45 @@ final class SerializerTest {
...
     
+    private struct NumberMessage: Codable {
+        let int: Int
+        let float: Float
+    }
+    
     @Test func test_decodeMessageWithNumbers() throws {
         let text = """
-        ["1","2", "topic","event",{"int":1,"float":1.1}]
+        ["1","2", "topic","event",{"int":1234,"float":1.1234}]
```
Comment (dsrees) [inline Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift]: This was one of my issues using JSONSerialization, the loss of some precision of float values when serializing back and forth between data and text, and why I went down the JsonElement route. 

I don't have an answer right now but curious if you have any ideas.

File: Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift [hunk 2]
```diff
@@ -91,37 +91,45 @@ final class SerializerTest {
...
+    private struct ValuePayload: Codable {
+        let value: String
     }
     
     @Test func test_decodeMessageWithoutJsonPayload() throws {
         // NOTE: Since NSNumber can be a float or an int, `1.0` will convert to `1`
         // but should be able to map `1` back into a float property.
         let text = """
-        ["1","2", "topic","event","payload"]
+        ["1","2", "topic","event",{"value": "payload"}]
```
Comment (dsrees) [inline Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift]: Just because the phoenix docs state that payload _should_ be a map, there is no requirement that it _has to be_ a map. We'll want to use `fragmentsAllowed` in our `options`.

The test values in the class were specifically chosen to handle edge cases and shouldn't be changed, otherwise we aren't guaranteeing that we are handling these edge cases.

**Review by dsrees — COMMENTED** [review]
I left a couple of contextual comments on the tests you changed, looking forward to hearing your thoughts.

Thanks for this, I probably won't move forward on this for a bit as I spend some time tinkering with some of the new ideas we've discussed.

File: Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift (hunk 1 above)
Comment (yeahphil) [reply Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift]: I changed these back. Definitely agree that the concern in round-trip encoding is that something changes!

But I'm not sure I understand the issue here -- do you have a test that shows loss of precision?

File: Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift (hunk 2 above)
Comment (yeahphil) [reply Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift]: Got it -- I added a couple commits that turn on `.fragmentsAllowed` and put the test values back as original.

**Comment by yeahphil** [comment]
Appreciate the review, and totally understand. Let me know if I can be of help. 

If you find any cases where this approach breaks as you're working, I would very much appreciate a heads-up.

File: Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift (hunk 1 above)
Comment (dsrees) [reply Tests/SwiftPhoenixClientTests/Serializer/SerializerTest.swift]: When the test case is 1.1, the decoded message payload decodes into a `float: 1.1000002`. This wasn't the case with JsonElement

Screenshot 2025-01-18 at 8 52 03 PM

Obviously this is an edge case and a near 30x speed improvement is worth the squeeze, but I'm just bringing context into problems I've come across in the wild and have tried to solve for in this 6.x pass

**Comment by yeahphil** [comment]
Gonna hoist this out of this comment thread: https://github.com/davidstump/SwiftPhoenixClient/pull/278#discussion_r1921319732

Ugh, float precision issues. On the one hand this is a fatal blow to the idea that you can roundtrip through the Foundation JSON encoding and trust that you're getting the same thing back. On the other hand, this is an issue in the current shipping version, and 1.1 can't be precisely represented in a floating point of any size -- that's just hidden most of the time. (Note that the test passes, and would provide the same precision as JSONElement if I had decoded into a Double rather than Float).

`JSONSerialization`, doesn't re-encode to the same thing though:
```swift
let decoded = try! JSONSerialization.jsonObject(with: "1.1".data(using: .utf8)!, options: .fragmentsAllowed)
let roundtripped = try! JSONSerialization.data(withJSONObject: decoded, options: .fragmentsAllowed)
print(String(data: roundtripped, encoding: .utf8)!) // 1.1000000000000001
```

Tons of detail on that issue here: https://github.com/swiftlang/swift-corelibs-foundation/issues/4255

Another thing is that `JSONElement` will exhibit this same behavior. They fixed the issue above, I believe when `JSONDecoder`/`JSONEncoder` got rewritten for iOS 17. But prior to that, those classes used `JSONSerialization` internally. So if you run tests on the 6.x branch under iOS 13-16 they will fail, and the re-encoded payload in `test_decodeMessageWithNumbers()` will be `{\"float\":1.1000000000000001,\"int\":1}`

I'm trying to think of the practical problem this creates -- probably representing exact decimal numbers, and not decoding to a floating point but to `Decimal` or some other exact representation. I don't think round trip encoding of any kind can guarantee that. Doing the partial decode -> generic CodableMessage approach I outlined earlier might be able to, but only when running on iOS 17 and up.

**Comment by dsrees** [comment]
@yeahphil a lot of good information here, thanks. 

Yeah I've been playing with your deferred codable message idea and think it might be a valuable solution. I still have some things I want to work through before moving forward with anything

**Comment by dsrees** [comment]
Closing in favor of #279. Thanks again for doing some of this benchmarking and proposing some alternatives.
